flexbox container leads to complete freeze of WebKit

nipunshukla21 opened the pull request:
#### fc6f33a5416ad3e3f003a734453f27308f927e1c
<pre>
flexbox container leads to complete freeze of WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=289072">https://bugs.webkit.org/show_bug.cgi?id=289072</a>
<a href="https://rdar.apple.com/146141220">rdar://146141220</a>

Reviewed by Sammy Gill.

In maybeCacheFlexItemMainIntrinsicSize, skip calling
updateBlockChildDirtyBitsBeforeLayout when we already
have a cached intrinsic size and the child doesn&apos;t need
layout for non-percentage reasons. This prevents
updateBlockChildDirtyBitsBeforeLayout from marking items
dirty solely due to percentage heights, which previously
caused each nested flex container to redundantly
re-layout all its descendants. Also skip setting
forceFlexItemRelayout to true unless the main axis size
has changed and we have already completed a layout pass
at the same container height.

A performance test is included to verify progression in
a pathological case with deep nesting. That microbenchmark
progresses from ~1.1 runs/s across 20 iterations to
~58 runs/s with this patch.

No functionality change should occur with this patch.

* PerformanceTests/Layout/nested-column-flexboxes-branching.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::performFlexLayout):
(WebCore::RenderFlexibleBox::maybeCacheFlexItemMainIntrinsicSize):
(WebCore::RenderFlexibleBox::layoutAndPlaceFlexItems):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: <a href="https://commits.webkit.org/309068@main">https://commits.webkit.org/309068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16fc8a8ac15a132b00583ce16cd9d01f377eb44a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/148935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/21648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/15217 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/157623 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/102365 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/150808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/22101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/21526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/114817 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/81756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/151895 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/17013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/133677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/95575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/16121 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/13979 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/5471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/125723 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/11598 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/160104 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/3095 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/13120 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/122872 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/21450 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/17994 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/123100 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/33540 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/21458 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/133392 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/77646 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/18388 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/10154 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/21060 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/84862 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/20792 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/20940 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/20848 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -534,6 +536,50 @@ RenderBlockRareData& RenderBlock::ensureBlockRareData()
     }).iterator->value;
 }
 
+LayoutUnit RenderBlock::cachedChildIntrinsicContentLogicalHeight(const RenderBox& child) const
```
This looks like a very generic caching setup. Does this apply to non-flex FCs? If not, is there a way to make it flex specific and not have it on RenderBlock?

nipunshukla21 replied:
I unfortunately don't have the best understanding of the mechanisms behind this (although I am reading the flexbox spec to try and improve that) but the gist of the issue is that in `RenderBlock::updateBlockChildDirtyBitsBeforeLayout()` and `updateFlexItemDirtyBitsBeforeLayout()` in `RenderFlexibleBox.cpp` as documented in the FIXMEs we relayout in cases we don't necessarily have to. For the test case below, which is linked in full [here](https://bugs.webkit.org/attachment.cgi?id=474512), this causes us to fail to load in any reasonable amount of time, whereas Chrome and Firefox do. Commenting out those code paths such that we don't check `flexItem.hasRelativeLogicalHeight()`/`childHasRelativeHeight()` does allow us to load.

This test case essentially creates a very wide tree of flex containers, 20 deep:
```
<style>.flex-container {
  flex-grow: 1;
  height: 100%;
  display: flex;
  flex-direction: column;
}
...
const createFlexContainerHtml = (layers, layer) => {
  if (layer > layers) {
    return ""
  }

  return `<div class="flex-container">
      <div>
        <div>
        ${createFlexContainerHtml(layers, ++layer)}
        </div>
      </div>
      <div>
        <div>
          ${createFlexContainerHtml(layers, ++layer)}
        </div>
      </div>
    </div>`
}

container.innerHTML = createFlexContainerHtml(20, 1)
```
The change here was an attempt to get EWS results based on my iteration upon an earlier patch [292314@main](https://commits.webkit.org/292314@main) which was reverted for causing [rdar://149102629](rdar://149102629). The wrinkle here is that patch doesn't fix the load hanging alone, as we still do a relayout in `RenderBlock` and I need to comment out that check as well. With all of that said, I believe that we do likely need some changes in RenderBlock itself, although they can almost definitely be more targeted than what I currently have. Do you have any advice or ideas on a better way to proceed?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2439,7 +2448,8 @@ void RenderFlexibleBox::layoutAndPlaceFlexItems(LayoutUnit& crossAxisOffset, Fle
         // width, so we need to compare to the size including the scrollbar.
         // FIXME: Should it include the scrollbar?
-        if (flexLayoutItem.flexedContentSize != mainAxisContentExtentForFlexItemIncludingScrollbar(flexItem))
+        bool mainAxisSizeChanged = flexLayoutItem.flexedContentSize != mainAxisContentExtentForFlexItemIncludingScrollbar(flexItem);
```
Do we really need `mainAxisSIzeChanged`? If this is true then we will enter the first branch. Otherwise we will enter the else branch and `mainAxisSizeChanged` has to be false right?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1860,19 +1864,26 @@ void RenderFlexibleBox::maybeCacheFlexItemMainIntrinsicSize(RenderBox& flexItem,
     // by definition we have an indefinite flex basis here and thus percentages should not resolve.
-    if (flexItem.needsLayout() || !m_intrinsicSizeAlongMainAxis.contains(flexItem)) {
+    if (flexItem.needsLayout() || !hasCachedIntrinsicSize) {
+        intrinsicSizeCache.add(flexItem);
```
I see that inside of `cacheFlexItemMainSize` we end up adding the flex item to `m_relaidOutFlexItems`. Is this something that we could potentially reuse since we seem to use it in a similar way? Although I'm not a fan of how we hold onto is as a member on `RenderFlexibleBox`, I'm thinking it might be a little confusing if we had two different structures that seem to be used in similar ways.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1466,9 +1468,11 @@ void RenderFlexibleBox::performFlexLayout(RelayoutChildren relayoutChildren)
+        layoutAndPlaceFlexItems(crossAxisOffset, lineItems, remainingFreeSpace, relayoutChildren, lineStates, gapBetweenItems, intrinsicSizeCache);
     }
 
+    m_logicalHeightAtLastPlacement = mainAxisContentExtent(LayoutUnit::max());
```
I think this needs a different name because this is not the logical height we are storing here.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2454,7 +2466,12 @@ void RenderFlexibleBox::layoutAndPlaceFlexItems(LayoutUnit& crossAxisOffset, Fle
+            // For column flex, the main axis is the block axis. If the main axis size
+            // didn't change AND this container has had a prior layout (not skipped due
+            // to content-visibility) at the same container height we can skip forced relayout
+            if (!isColumnFlow() || mainAxisSizeChanged || !everHadLayout() || wasSkippedDuringLastLayoutDueToContentVisibility().value_or(true) || mainAxisContentExtent(LayoutUnit::max()) != m_logicalHeightAtLastPlacement || intrinsicSizeCache.contains(flexLayoutItem.renderer))
```
We compute the line size at some point before we enter this function can we avoid recomputing it again?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2454,7 +2466,12 @@ void RenderFlexibleBox::layoutAndPlaceFlexItems(LayoutUnit& crossAxisOffset, Fle
+            // didn't change AND this container has had a prior layout (not skipped due
+            // to content-visibility) at the same container height we can skip forced relayout
+            if (!isColumnFlow() || mainAxisSizeChanged || !everHadLayout() || wasSkippedDuringLastLayoutDueToContentVisibility().value_or(true) || mainAxisContentExtent(LayoutUnit::max()) != m_logicalHeightAtLastPlacement || intrinsicSizeCache.contains(flexLayoutItem.renderer))
+                forceFlexItemRelayout = true;
```
Since this is basically a heuristic that we are using to try and determine when we can skip performing a layout to compute the cross size of flex items w/ percent height descendants we should see if we can find a way to express this in a way that improves readability.

You might be able to refactor this into a lambda or something with individual cases and then annotate each case with a reason *why* or *why not* that would require us to do layout or allow us to skip doing layout. That additional context is really important when it comes to optimizations like this imo and right now the comment just basically says the same thing as the code

alanbaradlay replied:
`// Our previous layout of the child was done without an override height set.`

This makes me think that layout did happen in this layout frame (flexing the flex item), so what makes 'wasSkippedDuringLastLayoutDueToContentVisibility' & co relevant here?

sammygill reviewed (commented):
Would you mind including the runs per second you got with and without your patch into the commit message?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1860,11 +1862,17 @@ void RenderFlexibleBox::maybeCacheFlexItemMainIntrinsicSize(RenderBox& flexItem,
+    //
+    // If we already have a cached intrinsic size and the child doesn't
+    // need layout for non-percentage reasons, we can skip updateBlockChildDirtyBitsBeforeLayout
+    // which would mark it dirty solely due to percentage heights.
```
How are we checking to see if we would need to do a layout for % height reasons here?

nipunshukla21 replied:
I updated the comment to better reflect why that if block works.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1466,9 +1466,11 @@ void RenderFlexibleBox::performFlexLayout(RelayoutChildren relayoutChildren)
+        layoutAndPlaceFlexItems(crossAxisOffset, lineItems, remainingFreeSpace, relayoutChildren, lineStates, gapBetweenItems, lineBreakLength);
     }
 
+    m_mainAxisContentExtentAtLastPlacement = lineBreakLength;
```
If we can, we should try to do this as close to the bottom of `performFlexLayout` as possible. Similarly. maybe this should be named "AtLastLayout" since last placement might be slightly ambiguous depending on whether you are before or after the call to `layoutAndPlaceFlexItems`?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2450,11 +2459,27 @@ void RenderFlexibleBox::layoutAndPlaceFlexItems(LayoutUnit& crossAxisOffset, Fle
-            // previous layout of the child was done without an override height set.
-            // So, redo it here.
-            forceFlexItemRelayout = true;
+            auto needsRelayoutForPercentHeightDescendants = [&] -> bool {
```
Maybe we could omit `-> bool` here

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.h:
```diff
@@ -315,6 +315,7 @@ class RenderFlexibleBox : public RenderBlock {
 
     LayoutUnit m_alignContentStartOverflow { 0 };
     LayoutUnit m_justifyContentStartOverflow { 0 };
+    LayoutUnit m_mainAxisContentExtentAtLastPlacement { 0 };
```
I think this could technically be `LayoutUnit m_mainAxisContentExtentAtLastPlacement;`

sammygill reviewed (commented):
You mention no behavior change with the patch in the commit message but maybe it's more accurate to say no functionality change since the way the code behaves definitely changed but the overall functionality (i.e. user facing effects) should remain the same?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.h:
```diff
@@ -315,6 +315,7 @@ class RenderFlexibleBox : public RenderBlock {
 
     LayoutUnit m_alignContentStartOverflow { 0 };
     LayoutUnit m_justifyContentStartOverflow { 0 };
+    LayoutUnit m_mainAxisContentExtentAtLastLayout;
```
Might be worth adding a comment saying this gets updated at the end of `performFlexLayout`

nipunshukla21 committed:
flexbox container leads to complete freeze of WebKit
https://bugs.webkit.org/show_bug.cgi?id=289072
rdar://146141220

Reviewed by Sammy Gill.

In maybeCacheFlexItemMainIntrinsicSize, skip calling
updateBlockChildDirtyBitsBeforeLayout when we already
have a cached intrinsic size and the child doesn't need
layout for non-percentage reasons. This prevents
updateBlockChildDirtyBitsBeforeLayout from marking items
dirty solely due to percentage heights, which previously
caused each nested flex container to redundantly
re-layout all its descendants. Also skip setting
forceFlexItemRelayout to true unless the main axis size
has changed and we have already completed a layout pass
at the same container height.

A performance test is included to verify progression in
a pathological case with deep nesting. That microbenchmark
progresses from ~1.1 runs/s across 20 iterations to
~58 runs/s with this patch.

No functionality change should occur with this patch.

* PerformanceTests/Layout/nested-column-flexboxes-branching.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::performFlexLayout):
(WebCore::RenderFlexibleBox::maybeCacheFlexItemMainIntrinsicSize):
(WebCore::RenderFlexibleBox::layoutAndPlaceFlexItems):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: https://commits.webkit.org/309068@main
